src/NelsonSiegelSvensson.jl

```diff
@@ -0,0 +1,179 @@
+abstract type ParametricModel end
+
+"""
+    NelsonSiegel(β₀, β₁, β₂, τ₁)
+
+Parameters of Nelson and Siegel (1987) parametric model. 
+
+# Examples
+
+```julia-repl
+julia> β₀, β₁, β₂, τ₁ = 0.6, -1.2, -1.9, 3.0
+julia> nsm = NelsonSiegelSvensson.NelsonSiegel.(β₀, β₁, β₂, τ₁)
+```
+"""
+struct NelsonSiegel <: ParametricModel
+    β₀
+    β₁
+    β₂
+    τ₁
+    #r² no need
+
+    #function NelsonSiegel(β₀, β₁, β₂, τ₁, r²)
+    function NelsonSiegel(β₀, β₁, β₂, τ₁)
+        #(τ₁ <= 0 || r² < 0 || r² > 1) && throw(DomainError("Wrong parameter ranges"))
+        (τ₁ <= 0) && throw(DomainError("Wrong parameter ranges"))
+        #return new(β₀, β₁, β₂, τ₁, r²)
+        return new(β₀, β₁, β₂, τ₁)
+    end
+end
+
+"""
+    NelsonSiegelSvensson(β₀, β₁, β₂, β₃, τ₁, τ₂)
+
+Parameters of Svensson (1994) parametric model. 
+
+# Examples
+
+```julia-repl
+julia> β₀, β₁, β₂, β₃, τ₁, τ₂ = 0.6, -1.2, -2.1, 3.0, 1.5
+julia> nssm = NelsonSiegelSvensson.NelsonSiegelSvensson.(β₀, β₁, β₂, β₃, τ₁, τ₂)
+```
+"""
+struct NelsonSiegelSvensson <: ParametricModel
+    β₀
+    β₁
+    β₂
+    β₃
+    τ₁
+    τ₂
+    #r² no need
+
+    #function NelsonSiegelSvensson(β₀, β₁, β₂, β₃, τ₁, τ₂, r²)
+    function NelsonSiegelSvensson(β₀, β₁, β₂, β₃, τ₁, τ₂)
+        #(τ₁ <= 0 || τ₂ <= 0 || r² < 0 || r² > 1) && throw(DomainError("Wrong parameter ranges"))
+        (τ₁ <= 0 || τ₂ <= 0) && throw(DomainError("Wrong parameter ranges"))
+        #return new(β₀, β₁, β₂, β₃, τ₁, τ₂, r²)
+        return new(β₀, β₁, β₂, β₃, τ₁, τ₂)
+    end
+end
+
+Base.zero(ns::NelsonSiegel, t) = ns.β₀ .+ ns.β₁ .* (1.0 .- exp.(-t ./ ns.τ₁)) ./ (t ./ ns.τ₁) .+ ns.β₂ .* ((1.0 .- exp.(-t ./ ns.τ₁)) ./ (t ./ ns.τ₁) .- exp.(-t ./ ns.τ₁))
+discount(ns::NelsonSiegel, t) = exp.(-t .* zero(ns, t))
+
+Base.zero(nss::NelsonSiegelSvensson, t) = nss.β₀ .+ nss.β₁ .* (1.0 .- exp.(-t ./ nss.τ₁)) ./ (t ./ nss.τ₁) .+ nss.β₂ .* ((1.0 .- exp.(-t ./ nss.τ₁)) ./ (t ./ nss.τ₁) .- exp.(-t ./ nss.τ₁)) .+ nss.β₃ .* ((1.0 .- exp.(-t ./ nss.τ₂)) ./ (t ./ nss.τ₂) .- exp.(-t ./ nss.τ₂))
+discount(nss::NelsonSiegel, t) = exp.(-t .* zero(nss, t))
+
+using LsqFit
+
+""" 
+    est_ns_params(swq::Vector{SwapQuote}, τₐₗₗ::Array{Float, 1})
+
+Return the NelsonSiegel fitted parameters.
+"""
+function est_ns_params(swq::Vector{SwapQuote}, τₐₗₗ::Array{Float64, 1} = [0.1, 0.15, 0.2, 0.3, 0.5, 0.75, 1, 1.5, 2, 3, 5, 7.5, 10]) where {Q<:ObservableQuote}
```

Comment:
This function ends up very similar to the one below. As a general rule, there should not be any code duplication. I think the best solution right now is to just keep the version below that takes yields and maturities as `AbstractArray`s. Then later, it would be easy to add a method that extracts interest rates from quotes. Also, using `SwapQuotes` in this case is a little dangerous, since swaps are usually quoted as par rates and not spot rates.